server/src/main/java/com/cloud/network/element/VirtualRouterElement.java

```diff
@@ -1223,8 +1223,7 @@ public boolean prepareMigration(final NicProfile nic, final Network network, fin
             }
         } else if (vm.getType() == VirtualMachine.Type.User) {
             assert vm instanceof UserVmVO;
-            final UserVmVO userVm = (UserVmVO) vm.getVirtualMachine();
```

Comment:
@nvazquez Have you observed similar issues with router VM using similar type casting in the if stmt above. eg: `DomainRouterVO router = (DomainRouterVO) vm.getVirtualMachine();`. Same casting is being used for router VM in other methods below.

Reply:
Thanks @sureshanaparti. I have not observed errors in those other castings but will be nice to solve the problem instead of the workaround I'm proposing in this fix, will take a look into that